### ISSUE TITLE
"uidmap" field has no "container" entry for user "demo1" (uid 1001)

### ISSUE THREAD

**Issue by bsidhu009** [nginx/unit#788]
I'm getting error "uidmap" field has no "container" entry for user "demo1" (uid 1001).

here my config. I not sure How to get this resolved. may be I need to create namespace. Is there any doc or script to create it?

```
{
    "listeners": {
        "*:80": {
            "pass": "applications/blog"
        }
    },
    "applications": {
        "blog": {
            "type": "php7.4",
            "group": "demo1",
            "user": "demo1",
            "root": "/home/demo1/public_html",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },

                "uidmap": [
                    {
                        "host": 1000,
                        "container": 0,
                        "size": 1000
                    }
                ],

                "gidmap": [
                    {
                        "host": 1000,
                        "container": 0,
                        "size": 1000
                    }
                ]
            }

        }
    }
}
```

**Comment by tippexs** [comment]
Hi @bsidhu009 sorry for late response and thanks for getting in touch? What OS do you use? Could you share that with us?

**Comment by tippexs** [comment]
Hi @bsidhu009 we did some debugging of this issue and found a bug in the way we are mapping the `host` or outer UIDs / GIDs with the `container` or inner UIDs / GIDs. Good news for you, ther is a configuration that will work for you.

```json
{
    "listeners": {
        "*:8999": {
            "pass": "applications/blog"
        }
    },
    "applications": {
        "blog": {
            "type": "php",
	    "user": "demo1",
	    "group": "demo1",
            "root": "/tmp/",
	    "index": "index.php",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },

                "uidmap": [
                    {
                        "host": 1001,
                        "container": 1001,
                        "size": 1
                    }
                ],

                "gidmap": [
                    {
                        "host": 1002,
                        "container": 1002,
                        "size": 1
                    }
                ]
            }

        }
    }
}

```

I will use the UID and GID from your application user `demo1` and map it exactly to the containers namespace. Make sure the IDs match what you can find in `/etc/passwd` or by issuing 

```shell
id demo1
uid=1001(demo1) gid=1002(demo1) groups=1002(demo1)
```
@alejandro-colomar will catch up here with a bug description. Thanks for reporting this to us.

**Comment by alejandro-colomar** [comment]
Hi!

This seems to be a bug.  I haven't yet looked at the code, but here's my rough guess:

(1)  Some unitd process that is in the host user namespace is getting the `uid` and `gid` of the user `demo1`.
(2)  Those numbers, which only have a meaning in the host user namespace are being sent through some inter-process communication to a process within the container user namespace, without being first mapped to the correct values within the container user namespace.
(3)  The inner process is interpreting them as if the values were the `uid` and `gid` in the container user namespace.  Of course, this is wrong.

That means that as long as the values in the host and container namespace match, it will work, which is why Timo has shown you an example where the mapping doesn't change uid or gid.

To confirm that this is the problem, we tried a few things, and we saw that with the mapping you were trying (extending a bit the size to make it work, for example to `2000`), the user that was being used for the app was the user 2001 in the host user namespace, which is wrong, but makes sense considering the bug.

To fix the bug, we'll need to do the translation before passing the value, if I can confirm that this is the exact bug that we're seeing.

**Comment by alejandro-colomar** [comment]
As a workaround, if it's good enough for you, you can have a mapping where the `"host"` and `"container"` values are the same (so the mapping is almost a no-op).  Of course, the `"size"` needs to be enough to cover the user/group that you're interested in.  Normally, a size of 1 should be enough.

**Comment by bsidhu009** [comment]
> Hi @bsidhu009 sorry for late response and thanks for getting in touch? What OS do you use? Could you share that with us?

Hi, I have updated OS version

**Comment by bsidhu009** [comment]
> Hi @bsidhu009 we did some debugging of this issue and found a bug in the way we are mapping the `host` or outer UIDs / GIDs with the `container` or inner UIDs / GIDs. Good news for you, ther is a configuration that will work for you.
> 
> ```json
> {
>     "listeners": {
>         "*:8999": {
>             "pass": "applications/blog"
>         }
>     },
>     "applications": {
>         "blog": {
>             "type": "php",
> 	    "user": "demo1",
> 	    "group": "demo1",
>             "root": "/tmp/",
> 	    "index": "index.php",
>             "isolation": {
>                 "namespaces": {
>                     "cgroup": true,
>                     "credential": true,
>                     "mount": true,
>                     "network": true,
>                     "pid": true,
>                     "uname": true
>                 },
> 
>                 "uidmap": [
>                     {
>                         "host": 1001,
>                         "container": 1001,
>                         "size": 1
>                     }
>                 ],
> 
>                 "gidmap": [
>                     {
>                         "host": 1002,
>                         "container": 1002,
>                         "size": 1
>                     }
>                 ]
>             }
> 
>         }
>     }
> }
> ```
> 
> I will use the UID and GID from your application user `demo1` and map it exactly to the containers namespace. Make sure the IDs match what you can find in `/etc/passwd` or by issuing
> 
> ```shell
> id demo1
> uid=1001(demo1) gid=1002(demo1) groups=1002(demo1)
> ```
> 
> @alejandro-colomar will catch up here with a bug description. Thanks for reporting this to us.

@tippexs config you have suggested is also not working, now error is 
"gidmap" field has missing suplementary gid mappings (found 1 out of 2).

**Comment by alejandro-colomar** [comment]
Maybe changing the `"size"` to something bigger than 1 fix the new issue?

**Comment by bsidhu009** [comment]
I have tried 2,5,7,10, still same issues ""gidmap" field has missing suplementary gid mappings (found 1 out of 2)."

**Comment by tippexs** [comment]
Whats the output of `id demo1`?

**Comment by bsidhu009** [comment]
uid=1002(demo1) gid=1002(demo1) groups=1002(demo1)
I had changed 1002 for both map as per above values

**Comment by alejandro-colomar** [comment]
I have some suspicion about the second problem:

In the tests we did, the demo1 user only had one group.  However, unitd tries to get the full list of groups to which a process belongs.  That means that if your `demo1` user belongs to several groups, all of them will need to have a valid mapping.

If that's the case, the workaround seems to be simple:  augment the mapping so that it covers all of those groups.  For that:

- As said before, the `"host"` and `"container"` values need to match.  This is to workaround the bug.  I hope to fix this soon.
- The uidmap should have a size of `1`.  I believe that should be enough.
- The uidmap should have a base value (both `"host"` and `"container"`) that matches the uid of `demo1`.
- The `"gidmap"` should have a base value (both `"host"` and `"container"`) that matches the `gid` that is lowest from the list of groups to which `demo1` belongs.
- The `"size"` of the `"gidmap"` should be enough to hold also the largest `gid` of all the list of groups.

Could you please test that?  If there's anything you don't understand, please ask, and I'll write some commands that you can run to help me debug your case (we couldn't reproduce so far the second problem, but I've been reading the code, and it seems what I just wrote; I hope you can confirm).

Thanks!

**Comment by alejandro-colomar** [comment]
Confirmed:

```json
{
    "listeners": {
        "*:8080": {
            "pass": "applications/hello-php"
        }
    },
    "applications": {
        "hello-php": {
            "type": "php",
            "group": "alx",
            "user": "alx",
            "root": "/home/alx/srv/www/",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },

                "uidmap": [
                    {
                        "host": 1000,
                        "container": 1000,
                        "size": 1
                    }
                ],

                "gidmap": [
                    {
                        "host": 112,
                        "container": 112,
                        "size": 1000
                    }
                ]
            }
        }
    }
}
```

```
2022/11/30 21:12:43.711 [debug] 28217#28217 about to get "alx" groups (uid:1000, base
 gid:1000)
2022/11/30 21:12:43.711 [debug] 28217#28217 user "alx" has gids:1000,24,25,29,30,44,4
6,102,108,112,115^@
2022/11/30 21:12:43.711 [error] 28217#28217 "gidmap" field has missing suplementary gid mappings (found 3 out of 11).
```


So, groups `112`, `115`, and `1000` are covered by the map.  The other 8 out of 11 are not mapped, and so the app can't be run.

The good news is that the workaround is easy.  Write a map that covers all of them.

I'm working on a fix.

**Comment by surjit** [comment]
Hello @alejandro-colomar, help me with commands How to debug.

**Comment by alejandro-colomar** [comment]
@artemkonev 

The documentation doesn't specify that unitd(8) needs to be able to map all groups of the user (and the kernel doesn't need it either).  The kernel can work with incomplete mappings, and the unmapped groups will just be translated to the overflow gid (65334):

user_namespaces(7):
```
   Unmapped user and group IDs
       There are various places where an unmapped user ID (group  ID)  may  be
       exposed  to  user  space.  For example, the first process in a new user
       namespace may call getuid(2) before a user ID mapping has been  defined
       for  the  namespace.   In  most such cases, an unmapped user ID is con‐
       verted to the overflow user ID (group ID); the default  value  for  the
       overflow  user  ID  (group  ID)  is  65534.   See  the  descriptions of
       /proc/sys/kernel/overflowuid   and   /proc/sys/kernel/overflowgid    in
       proc(5).
```

So, if we need to get all of those groups, I'd document that the mapping needs to cover all of them.  Otherwise, we could just change unit to work with the overflow gid.

CC @i4ki

**Comment by alejandro-colomar** [comment]
@surjit

Please run `id demo1` and paste what it prints.

**Comment by surjit** [comment]
@alejandro-colomar, I didn't understand math of "That 977 is: 1000 - 24 + 1."

**Comment by alejandro-colomar** [comment]
> @alejandro-colomar, I didn't understand math of "That 977 is: 1000 - 24 + 1."

From here:

```
$ id alx
uid=1000(alx) gid=1000(alx) groups=1000(alx),24(cdrom),25(floppy),29(audio),30(dip),44(video),46(plugdev),102(netdev),108(bluetooth),112(lpadmin),115(scanner)
```
You can see that my user belongs to multiple groups: `24, 25, 29, 30, 44, 46, 102, 108, 112, 115, and 1000`.  So I need a mapping that maps the range `24 ... 1000`.  Since `gidmap` needs the start gid and the size, the start of the range is obviously `24`, and the range is something that will include `1000`.

So, if the start (`"host"`) is `24`, and the size is `1`, the range will only be `24`.  If the size is `2`, the range will be `24, 25`. ... inow we can se a pattern: `size == last - first + 1`.  If you substitute in there my values, you'll get that for a range `24 ... 1000`, the first is `24`, the last is `1000`, and the size results in `977`.

I hope this clarifies.

**Comment by artemkonev** [comment]
@alejandro-colomar noted.

**Comment by alejandro-colomar** [comment]
> @artemkonev
> 
> The documentation doesn't specify that unitd(8) needs to be able to map all groups of the user (and the kernel doesn't need it either). The kernel can work with incomplete mappings, and the unmapped groups will just be translated to the overflow gid (65334):
> 
> user_namespaces(7):
> 
> ```
>    Unmapped user and group IDs
>        There are various places where an unmapped user ID (group  ID)  may  be
>        exposed  to  user  space.  For example, the first process in a new user
>        namespace may call getuid(2) before a user ID mapping has been  defined
>        for  the  namespace.   In  most such cases, an unmapped user ID is con‐
>        verted to the overflow user ID (group ID); the default  value  for  the
>        overflow  user  ID  (group  ID)  is  65534.   See  the  descriptions of
>        /proc/sys/kernel/overflowuid   and   /proc/sys/kernel/overflowgid    in
>        proc(5).
> ```
> 
> So, if we need to get all of those groups, I'd document that the mapping needs to cover all of them. Otherwise, we could just change unit to work with the overflow gid.
> 
> If you need me to detail the problem in more depth, we can talk about it.
> 
> CC @i4ki

Self correction.  I had forgotten about this issue lately, but a patch to the man-pages made me read it again...

```
       The /proc/ pid /setgroups file was added in Linux 3.19, but  was  back‐
       ported to many earlier stable kernel series, because it addresses a se‐
       curity  issue.   The  issue  concerned  files  with permissions such as
       "rwx---rwx".  Such files give fewer permissions to "group" than they do
       to "other".  This means that dropping groups using  setgroups(2)  might
       allow  a process file access that it did not formerly have.  Before the
       existence of user namespaces this was not a concern, since only a priv‐
       ileged process (one with the CAP_SETGID  capability)  could  call  set‐
       groups(2).   However,  with the introduction of user namespaces, it be‐
       came possible for an unprivileged process to create a new namespace  in
       which the user had all privileges.  This then allowed formerly unprivi‐
       leged  users to drop groups and thus gain file access that they did not
       previously have.  The /proc/pid/setgroups file  was  added  to  address
       this security issue, by denying any pathway for an unprivileged process
       to drop groups with setgroups(2).
```

So, dropping groups is an issue, and that's why unmapped groups error out in the kernel, because otherwise it would be a security concern.  So we need to be able to map all users and groups, and probably document something in the passage that documents gid_map, so that users understand this error when it happens.

TL;DR: All groups must be mapped.

This is orthogonal to fixing the bug, though.  I still struggle finding the fix.

**Comment by surjit** [comment]
Error :- "gidmap" field has missing suplementary gid mappings (found 1 out of 2).
uid=1001(demo1) gid=1001(demo1) groups=1001(demo1),998(nobody)

```
{
    "listeners": {
        "*:8082": {
            "pass": "applications/blog"
        }
    },
    "applications": {
        "blog": {
            "type": "php",
            "group": "demo1",
            "user": "demo1",
            "root": "/home/demo1/public_html",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },
                "uidmap": [
                    {
                        "host": 1001,
                        "container": 1001,
                        "size": 1
                    }
                ],
                "gidmap": [
                    {
                        "host": 1001,
                        "container": 1001,
                        "size": 1
                    }
                ]
            }
        }
    }
}
```

below config works
uid=1001(demo1) gid=1001(demo1) groups=1001(demo1),998(nobody)
```
{
    "listeners": {
        "*:8082": {
            "pass": "applications/blog"
        }
    },
    "applications": {
        "blog": {
            "type": "php",
            "group": "demo1",
            "user": "demo1",
            "root": "/home/demo1/public_html",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },
                "uidmap": [
                    {
                        "host": 1001,
                        "container": 1001,
                        "size": 1
                    }
                ],
                "gidmap": [
                    {
                        "host": 998,
                        "container": 998,
                        "size": 4
                    }
                ]
            }
        }
    }
}
```

**Comment by i4ki** [comment]
Sorry for the late reply...

The problem is that the user created a mapping for  (0-999 in the container) -> (1000-1999 in the host) and provided `user: demo1` which resolved to `1001` in the host and this number is outside the available range in the container.

If there's no `isolation.rootfs` involved, then it should error and we should not map any additional user into the namespace...

The issue here is that OP provided `isolation.namespaces.credential = true` then whatever ids Unit cached for `user` and `group` are meaningless in the container, and yes this was a known issue but with planned solutions. The first implementation of credential namespace had the (undocumented) limitation that `user` and `group` would only work with creds ns *if* the mappings for the container matched the host.

The `user`, `group`, `isolation.namespace.credential` and `rootfs` options must work together to provide a good user experience (ala LXC, docker, runc, etc).

So when we introduced the `rootfs` option, a lot of improvements were left out but were in the roadmap (mounting tmpfs/procfs (implemented), mounting arbitrary volumes, etc) and one of those planned improvements was making the `user` and `group` option be relative to the `rootfs` (if set). Our goal was to make it possible to use `docker` (or similar) container images with Unit, then the `user` and `group` provided in the configuration would need to map what's configured in the container image (in simple terms, what's defined in its `/etc/passwd`, `/etc/groups`). We should cache the uids/gids in the bootstrap *only for host apps*, which means apps with no isolation. It needs to be cached for the host for several reasons, but mostly to integrate with remote state (Active Directory, etc) but for isolated apps it needs to fetch/parse the information from the image, just before `setuid()`.

Then, IMHO, if `credential` namespace is `true` and `user`/`group` is set, then the correct fix here would be:

- load uids/gids from the container filesystem (if `rootfs` is not set, the container filesystem is the host filesystem, and this matches the behavior today)
- if `rootfs` is set but no `<rootfs>/etc/passwd`, etc, is in place, then fail as isolated environment knows nothing about those provided user and group.
- BONUS: provide a `uid` and `gid` option for users using custom rootfs (no real OS images), so they explicitly say the uid and gid of the container.

Then, if OP has `demo` and `demo1` in the host as uid `1000` and `1001`, then for the mapping below:
```
{
    "host": 1000,
    "container": 1000,
    "size": 2
}
```
One option would be to just drop the file below inside the `rootfs`
```
# users needed by the app
demo:x:1000:1000::/home/demo:/bin/bash
demo1:x:1001:1001::/home/demo1:/bin/bash
```
In the future where this is implemented, users can pack their applications inside docker containers defined by Dockerfiles, which creates and setup the image with commands (`useradd`, `groupadd`, etc).

For simpler use cases, as is the case of the OP (no rootfs), the configuration below would suffice (the BONUS):

```
{
   "type": "php",
   "uid": 5000,  // any number defined in the uidmap
   "gid": 5000,  // any number defined in the gidmap
   "isolation": {
       "namespaces": {
           "namespace": { "credential": true }
       },
       "uidmap": [
           {"host": 1000, "container": 5000, "size": 2}
       ],
       "gidmap": [
           {"host": 1000, "container": 5000, "size": 2}
       ]
   }
}
```

The thing said about supplementary groups is a separate issue (I think) and the current implementation also has flaws yet to be addressed, for better security, ideally, there should be a way of dropping all supplementary groups of the host, inside the container. (But I never thought much about this)

**Comment by alejandro-colomar** [comment]
Hi Thiago!

> Sorry for the late reply...

No problem. :)

> 
> The problem is that the user created a mapping for (0-999 in the container) -> (1000-1999 in the host) and provided `user: demo1` which resolved to `1001` in the host and this number is outside the available range in the container.
> 
> If there's no `isolation.rootfs` involved, then it should error and we should not map any additional user into the namespace...
> 
> The issue here is that OP provided `isolation.namespaces.credential = true` then whatever ids Unit cached for `user` and `group` are meaningless in the container, and yes this was a known issue but with planned solutions. The first implementation of credential namespace had the (undocumented) limitation that `user` and `group` would only work with creds ns _if_ the mappings for the container matched the host.
> 
> The `user`, `group`, `isolation.namespace.credential` and `rootfs` options must work together to provide a good user experience (ala LXC, docker, runc, etc).
> 
> So when we introduced the `rootfs` option, a lot of improvements were left out but were in the roadmap (mounting tmpfs/procfs (implemented), mounting arbitrary volumes, etc) and one of those planned improvements was making the `user` and `group` option be relative to the `rootfs` (if set). Our goal was to make it possible to use `docker` (or similar) container images with Unit, then the `user` and `group` provided in the configuration would need to map what's configured in the container image (in simple terms, what's defined in its `/etc/passwd`, `/etc/groups`). We should cache the uids/gids in the bootstrap _only for host apps_, which means apps with no isolation. It needs to be cached for the host for several reasons, but mostly to integrate with remote state (Active Directory, etc) but for isolated apps it needs to fetch/parse the information from the image, just before `setuid()`.
> 
> Then, IMHO, if `credential` namespace is `true` and `user`/`group` is set, then the correct fix here would be:
> 
>     * load uids/gids from the container filesystem (if `rootfs` is not set, the container filesystem is the host filesystem, and this matches the behavior today)

Hmmm, this explains why the mappings were reversed.  But then we need to somehow map the uids/gids that we got from the host filesystem's `/etc/passwd` to the container uids/gids _before_ we treat them as container uids/gids.  That mapping was missing, and is what I fixed in <https://github.com/nginx/unit/pull/800>.  Please review that PR.

> 
>     * if `rootfs` is set but no `<rootfs>/etc/passwd`, etc, is in place, then fail as isolated environment knows nothing about those provided user and group.

This makes sense.  So when we do that (AFAIU, we don't support this yet), we'll need to add a conditional to only perform the mapping from host to container during the caching, so in case `rootfs` in not set, we don't map before caching.

> 
>     * BONUS: provide a `uid` and `gid` option for users using custom rootfs (no real OS images), so they explicitly say the uid and gid of the container.

Yes, I missed this feature.

> 
> 
> Then, if OP has `demo` and `demo1` in the host as uid `1000` and `1001`, then for the mapping below:
> 
> ```
> {
>     "host": 1000,
>     "container": 1000,
>     "size": 2
> }
> ```
> 
> One option would be to just drop the file below inside the `rootfs`
> 
> ```
> # users needed by the app
> demo:x:1000:1000::/home/demo:/bin/bash
> demo1:x:1001:1001::/home/demo1:/bin/bash
> ```
> 
> In the future where this is implemented, users can pack their applications inside docker containers defined by Dockerfiles, which creates and setup the image with commands (`useradd`, `groupadd`, etc).
> 
> For simpler use cases, as is the case of the OP (no rootfs), the configuration below would suffice (the BONUS):
> 
> ```
> {
>    "type": "php",
>    "uid": 5000,  // any number defined in the uidmap
>    "gid": 5000,  // any number defined in the gidmap
>    "isolation": {
>        "namespaces": {
>            "namespace": { "credential": true }
>        },
>        "uidmap": [
>            {"host": 1000, "container": 5000, "size": 2}
>        ],
>        "gidmap": [
>            {"host": 1000, "container": 5000, "size": 2}
>        ]
>    }
> }
> ```
> 
> The thing said about supplementary groups is a separate issue (I think) and the current implementation also has flaws yet to be addressed, for better security, ideally, there should be a way of dropping all supplementary groups of the host, inside the container. (But I never thought much about this)

Dropping groups is a security issue.  See `user_namespaces(7)`:

```
       The /proc/ pid /setgroups file was added in Linux 3.19, but  was  back‐
       ported to many earlier stable kernel series, because it addresses a se‐
       curity  issue.   The  issue  concerned  files  with permissions such as
       "rwx---rwx".  Such files give fewer permissions to "group" than they do
       to "other".  This means that dropping groups using  setgroups(2)  might
       allow  a process file access that it did not formerly have.  Before the
       existence of user namespaces this was not a concern, since only a priv‐
       ileged process (one with the CAP_SETGID  capability)  could  call  set‐
       groups(2).   However,  with the introduction of user namespaces, it be‐
       came possible for an unprivileged process to create a new namespace  in
       which the user had all privileges.  This then allowed formerly unprivi‐
       leged  users to drop groups and thus gain file access that they did not
       previously have.  The /proc/pid/setgroups file  was  added  to  address
       this security issue, by denying any pathway for an unprivileged process
       to drop groups with setgroups(2).
```

Thanks!  And happy new year :)

Alex

**Comment by artemkonev** [comment]
@bsidhu009 Have you tried the updated Unit version? I also draw your attention to the fact that `/home` directories may have limited permissions and therefore not be suitable for such tests (`/tmp` is also a bad choice).
